### ISSUE TITLE
Keyboard Navigation for React Grid

### ISSUE THREAD

**Issue by LazyLahtak** [DevExpress/devextreme-reactive#3141]
# The Problem

[React Grid](https://devexpress.github.io/devextreme-reactive/react/grid/) users can use the keyboard to focus and control interactive elements (filters, buttons, editors). The Grid, however, does not provide full keyboard navigation support. This shortcoming, particularly evident in such use cases as [Inline Cell Editing](https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/editing/#inline-cell-editing), decreases the Grid's overall accessibility.

# The Proposed Solution

A new `TableKeyboardNavigation` plugin allows you to enable built-in keyboard navigation.

## Controlled Mode

```js
const [focusedCell, setFocusedCell] = useState(null);
const {
  rowKey,         // A unique key that identifies a row; formatted as "RowType_RowId"
  columnKey       // A unique key that identifies a column; formatted as "ColumnType_ColumnName"
} = focusedCell;
...
<TableKeyboardNavigation
  focusedCell={focusedCell}
  onFocusedCellChange={setFocusedCell}
/>
```

## Uncontrolled Mode

```js
const defaultFocusedRowId = 2;
const defaultFocusedColumn = "name";
const defaultFocusedCell = {
  rowKey: `${Table.ROW_TYPE}_${defaultFocusedRowId}`,
  columnKey: `${Table.COLUMN_TYPE}_${defaultFocusedColumn}`,
}
...
<TableKeyboardNavigation
  defaultFocusedCell={focusedCell}
  focusedRowEnabled // Applies a focused style to the row that contains the focused cell
/>
```

## Excel-like Navigation

The following keys can be used:

- Tab / Shift + Tab - Moves the focus one cell forward/back. 
- Arrow keys - Navigate to the corresponding element (above, below, left, right).
- Ctrl (Cmd) + Up/Down Arrow - Moves the focus between the header row, filter row, table, and footer. 
- Ctrl (Cmd) + F - Moves the focus to the search panel (if it is visible).

## Actions on Interactive Elements

The following keys can be used:

- Enter (Return) - Performs the action of the focused element (expands/collapses row or group, switches the focused cell to the edit state). 
- Esc - Discards edits (if a row is in the edit state).
- Space - Selects/deselects the focused row (if selection is enabled).
- Ctrl (Cmd) + Left/Right Arrow - Collapses/expands the focused row in [tree mode](https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/tree-data/).

# We Need Your Feedback

## Take a Quick Poll

[Do you find the keyboard navigation enhancements useful?](https://forms.gle/2EeFGxutUvKHSBnm7)

# Get Notified of Updates

Subscribe to this thread or our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by jesusgp22** [comment]
looks good but I'm also expecting some controls to toggle expanded rows, using Ctrl to do so recursively would be best

**Comment by LazyLahtak** [comment]
Hi @jesusgp22,

Thank you for your response. We don't plan to support built-in recursive expand. I suggest that you implement this scenario using [tree data in controlled mode](https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/tree-data/#controlled-mode). Here is an [example](https://codesandbox.io/s/fast-frog-cepgn?file=/demo.js) that illustrates how to do this. You can use it with Keyboard Navigation as well.

**Comment by las3r** [comment]
This seems like a good starting point. I would however ask you guys to also include shift-selection of ranges in the grid as an extension to this.

**Comment by michaeltnguyen** [comment]
I think this feature will be a great addition.  Do you have any experimental builds with it, or do you have a (very rough) ETA?  We're trying to decide if we can wait for this feature, or if we should try to roll our own for the time being.

**Comment by LazyLahtak** [comment]
Thank you for your feedback.

Now we are preparing the next major release to support React 17. For this, we need to fix some issues (#2728 and #2727). Also we are still waiting for the official [enzyme adapter](https://github.com/enzymejs/enzyme/issues/2429) for React 17 to be able to run our tests.
After that, we will continue to develop this feature. We will update this thread once we have any news.

**Comment by mklosterboer** [comment]
DevExpress Javascript has posted a new [2021 roadmap](https://community.devexpress.com/blogs/javascript/archive/2021/02/24/devextreme-javascript-2021-roadmap.aspx).
This roadmap says the following of DevExtreme Reactive:
>We expect to support our Reactive components and to introduce minor enhancements in 2021. We don’t plan on any major updates to this product line in 2021.

With that in mind, is this feature still going to be developed or will efforts be moving to the DevExpress Javascript components outlined in this new roadmap?

My team has committed pretty heavily to this package and this is a highly desirable feature for us and, by extension, our customers.

**Comment by woodreamz** [comment]
> DevExpress Javascript has posted a new [2021 roadmap](https://community.devexpress.com/blogs/javascript/archive/2021/02/24/devextreme-javascript-2021-roadmap.aspx).
> This roadmap says the following of DevExtreme Reactive:
> 
> > We expect to support our Reactive components and to introduce minor enhancements in 2021. We don’t plan on any major updates to this product line in 2021.
> 
> With that in mind, is this feature still going to be developed or will efforts be moving to the DevExpress Javascript components outlined in this new roadmap?
> 
> My team has committed pretty heavily to this package and this is a highly desirable feature for us and, by extension, our customers.

I have the same concern, the Grid is an important component for our webapps and we spent a lot of time on this component. We were waiting for new features and enhancements and the new roadmap worries me about the futur of the DevExtreme Reactive package.

**Comment by MaximKudriavtsev** [comment]
Hi @mklosterboer, @woodreamz.

We will release the Keyboard Navigation feature as promised. Also, we won't stop maintaining the DevExtreme Reactive package until we have a better substitution and a smooth migration path.

**Comment by rrajaste** [comment]
So there is currently no ETA on this feature? Not even very rough estimate? Will this be available within 2021?

**Comment by Krijovnick** [comment]
Hi @rrajaste

We are working on this. We will update this issue once we have any news to share.

**Comment by shrikantpandeydotnet** [comment]
Hi Team, we are looking for similar navigation ability like this https://5vy2q8owj4.csb.app/
 and it would be key for us in deciding the solution we pick. I think this solution discuss in this thread would give us same sort of result. Could you please share us any update on this feature ?

**Comment by zdzarsky** [comment]
Honestly what I love over DXGrid solution is a flexibility, so having a controlled mode is a must for us.
We would like to employ our own key-shortcuts using some existing well tested solutions, so having the control of edited cell depending on row and column is a wonderful idea an I cheer it to happen.

**Comment by woodreamz** [comment]
Any update on this feature? I don't want to be impatient but as I said in a previous comment , I am worried about the future of this package. It seems to be in "maintenance" and according to the changelogs, no new features have been released in the past year (last one was on 2020-07-17).

**Comment by Krijovnick** [comment]
@woodreamz 
We did not plan any features to the reactive line in 2021. This you can find in our [roadmap 2021](https://community.devexpress.com/blogs/javascript/archive/2021/02/24/devextreme-javascript-2021-roadmap.aspx) in `DevExtreme Reactive` section. 
But since we promised the Keyboard Navigation feature a long time ago, we decided to add it to our plans on 2021 and are currently working on it. Don't worry, we will release the Keyboard Navigation feature as promised as soon as possible.

**Comment by zdzarsky** [comment]
Guys, how is the progress. This feature is quite needed by our team, do you have a progress tracking anywhere?

**Comment by woodreamz** [comment]
> Guys, how is the progress. This feature is quite needed by our team, do you have a progress tracking anywhere?

It seems there is a PR in progress: #3386

**Comment by zdzarsky** [comment]
It seems it has been merged to master! 🎉 
@LazyLahtak when are you planning the next release?

**Comment by Krijovnick** [comment]
We have released the [3.0.0-beta.1](https://github.com/DevExpress/devextreme-reactive/blob/master/CHANGELOG.md#300-beta1-2021-10-21) version of DevExtreme Reactive Grid with Keyboard Navigation. Please check it.
